### DISCUSSION TITLE
Typescript Problems

### DISCUSSION THREAD

**Discussion by thejozika** [react-hook-form/react-hook-form#11740] [Q&A]
With the following code i get the following typescript error for name.
But it should work because a Record should be contained?

TS2345: Argument of type Key is not assignable to parameter of type Path<F>
Type string is not assignable to type Path<F>

The error suggests, that key is just treated as a string...

```
import {UseFormRegisterReturn, UseFormSetValue } from "react-hook-form";


interface Props<Key extends string , Val, F extends Record<Key, Val[]>>{
    value:Val[]
    register:UseFormRegisterReturn<Key>
    setValue:UseFormSetValue<F>
}

const Input = <Key extends string, Val, F extends Record<Key, Val[]>>(
    {register,setValue,value}:Props<Key, Val,F>) => {
    const {name } = register
    setValue(name, value)
    return <div>empty</div>
}


export default Input
```

**Comment by sujinleeme** [comment]
If you use `useFormContext`, you can avoid prop drilling. 

You can use `UseFormReturn<FormValues>` type consistently to avoid type confliction and decorations.

[[Playground]](https://www.typescriptlang.org/play?pretty=true#code/JYWwDg9gTgLgBAJQKYEMDG8BmUIjgcilQ3wG4AoUSWOAbzgFUBnJAMWhGRgFcoA7ADRxuLdlDwBfONlwEi6GAFoAFhAgBrRZg5ly5GAE8wSOGJAA1FABtuSJnAC8dcnDggkIAEZIoALjhMMFDAfADmFK6hONxg-oHBYRQSFPpGJmaOjKIcXLx8ADxmljZ2AHwpITA+mOgmAAo4YEyFwEhWACYAcijucEgAHlV87fbqSAYQmKYcxbZMpc6uRKHAgT7+ZgDahEgra1D4ALoRAUgws0gbHNss59a2RyeYrR3d7hsvXT1IJwBu95dpuILkxNqxPm8kMdyBI9GgIHxAnAAJJ8MDceBOFptL69AZDEZwMYTKZFAHzAAUtBcpzuJQEMP8DQgTWxr2+pQAlI4FtSlmc8nB8u1gL9Sh4wIZ8gB6EVipIpeGI+D0W4XITLVZVKBwKROERsDiFGbk0oUzmKhFIgCyHm8UFR6MxcApNPyjox+Xw7i8PnwpRp-L22octE1+1hrlczxxkIc3rtfsDcH+JXjAGVuKEUAdk2qAaH8yVI3BpQGLXCrfAAOLRMAe52u1zutGe-BRCAxf3J8MhsO7LU+EvRiHfeMdrvJ1O2eMAIQAKun8HmzhdC6uASWy+QK5XlXBZyh2g2AKL9HpgKxIACMmSbQobXr4EEUxP9pelcHnaT6UBwfjgGBlBMGN2V6VY4GfeABkHdoewHfZQ17Idk1A3EkHHOtlyjFMC3wBclxXOkZ1oItbC3ctLX3Q9j1bGAzwvK8ACY7zdR92ywgMcOQqAkIQ7Vh2kUd3EwzswGwqNpww2gAGYpGlT9v2MX9-38edgNwko4AgqCAiCEJQiItdSI3YsaW3CsgA)


```ts
import { UseFormReturn, useForm } from 'react-hook-form';

type FormValues = {
  member: string;
  group: string;
};

type Form = UseFormReturn<FormValues>;

interface Props<FieldName extends keyof FormValues> {
  register: Form['register'];
  setValue: Form['setValue'];
  fieldName: FieldName;
  value: FormValues[FieldName];
}

const Input = <FieldName extends keyof FormValues>({
  setValue,
}: Props<FieldName>) => {
  return <div>empty</div>;
};

const { setValue, register } = useForm<FormValues>();

const MemberInput = (
  <Input<'member'>
    register={register}
    fieldName='member'
    value='Sugar'
    setValue={setValue}
  />
);

const GroupInput = (
  <Input<'group'>
    register={register}
    fieldName='group'
    value='BTS'
    setValue={setValue}
  />
);


const BadInputExample1 = (
  <Input<'no-key'> // Type error: the fieldName is not existed
    register={register}
    fieldName='group'
    value='BTS'
    setValue={setValue}
  />
);

const BadInputExample2 = (
  <Input<'group'>
    register={register}
    fieldName='group'
    value={3} // Type error: The value is not string
    setValue={setValue}
  />
);
```
